chop/penalties.py

```diff
@@ -0,0 +1,80 @@
+"""
+This module contains various penalties / regularizers.
+They function batch-wise, similar to objects in `chop.constraints`.
+
+Code inspired from https://github.com/openopt/copt/
+"""
+
+import torch
+import torch.nn.functional as F
+
+from chop import utils
+
+
+class L1:
+    """L1 penalty. Batch-wise function."""
+
+    def __init__(self, alpha):
+        if alpha < 0:
+            raise ValueError("alpha must be non negative.")
+        self.alpha = alpha
+
+    def __call__(self, x):
+        batch_size = x.size(0)
+        return abs(x.view(batch_size, -1)).sum(dim=-1)
+
+    def prox(self, x, step_size=None):
+        """Prox operator for L1 norm. This is given by soft-thresholding."""
+        return torch.sign(x) * F.relu(abs(x) - self.alpha * step_size)
+
+
+class GroupL1:
+    """
+    Group LASSO penalty. Batch-wise function.
+    """
```

Comment:
add a description of the parameters in the init (specially groups since its non-trivial)

Reply:
even better: add an example. I'm reading the tests and its non-trivial to understand how the groups are represented

Reply:
Working on adding an adversarial-patch image example. I'll keep it in this branch.

The groups are an iterable of iterables. They're transformed in an iterable of torch.Tensors used for indexing. I can't directly use a torch.Tensor of higher dim because the groups might be of different sizes.

Reply:
Added explanations in the docstring.